third_party/roughenough/src/client.rs

```diff
@@ -0,0 +1,295 @@
+// Copyright 2017-2019 int08h LLC
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::{
+    merkle::root_from_paths, sign::Verifier, RtMessage, Tag, CERTIFICATE_CONTEXT,
+    SIGNED_RESPONSE_CONTEXT,
+};
+use byteorder::{LittleEndian, ReadBytesExt};
+use ring::{rand, rand::SecureRandom as _};
+use std::collections::HashMap;
+
+/// Creates a 64 byte nonce.
+pub fn create_nonce() -> [u8; 64] {
+    let rng = rand::SystemRandom::new();
+    let mut nonce = [0u8; 64];
+    rng.fill(&mut nonce).unwrap();
+
+    nonce
+}
+
+/// Converts a nonce to a Roughtime request.
+pub fn make_request(nonce: &[u8]) -> Vec<u8> {
+    let mut msg = RtMessage::new(1);
+    msg.add_field(Tag::NONC, nonce).unwrap();
+    msg.pad_to_kilobyte();
+
+    msg.encode().unwrap()
+}
+
+/// The parsed data extracted from a Roughtime response.
+pub struct ParsedResponse {
+    pub verified: bool,
+    pub midpoint: u64,
+    pub radius: u32,
+}
+
+/// Decodes, parses and validates Roughtime responses.
+pub struct ResponseHandler {
+    pub_key: Option<Vec<u8>>,
+    msg: HashMap<Tag, Vec<u8>>,
+    srep: HashMap<Tag, Vec<u8>>,
+    cert: HashMap<Tag, Vec<u8>>,
+    dele: HashMap<Tag, Vec<u8>>,
+    nonce: [u8; 64],
+}
+
+impl ResponseHandler {
+    pub fn new(pub_key: Option<Vec<u8>>, response: RtMessage, nonce: [u8; 64]) -> ResponseHandler {
+        let msg = response.into_hash_map();
+        let srep = RtMessage::from_bytes(&msg[&Tag::SREP])
+            .unwrap()
+            .into_hash_map();
+        let cert = RtMessage::from_bytes(&msg[&Tag::CERT])
+            .unwrap()
+            .into_hash_map();
+        let dele = RtMessage::from_bytes(&cert[&Tag::DELE])
+            .unwrap()
+            .into_hash_map();
+
+        ResponseHandler {
+            pub_key,
+            msg,
+            srep,
+            cert,
+            dele,
+            nonce,
+        }
+    }
+
+    pub fn extract_time(&self) -> ParsedResponse {
```

Comment:
Not for this PR, but it might be nice to make the error handling here more Rust-like (e.g. have this method `-> Result<ParsedResponse, Error>` and then have all of the validate checks return `-> Result<(), Error>` rather than `-> bool`,  and use `?` (or some [conversion](http://tinyurl.com/rust-transform)) rather than `.unwrap()`)